### PR TITLE
feat: picture in picture calling PoC [WPB-7221]

### PR DESCRIPTION
**Pull request by PatrykBuniX**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-7221" title="WPB-7221" target="_blank">SpikeWPB-7221</a>  [Web] Investigate and Create POC for Picture-in-Picture Calling UI pop-out
  </td></table>
  <br />
 

## Description

The POC for Picture-in-Picture Calling UI pop-out.

There's a new calling view mode added - `DETACHED_WINDOW`, when enabled, calling ui will be gone from the webapp and new window with the calling ui will be displayed instead.

Currently, for PoC there's no support for:
- call participant list actions
- choosing a screen to share

as those need some more (product/design) investigation (where those items should be put in the PiP view).

For browsers that do not support opening an external window, the calling ui will be opened in the new browser tab instead.

The feature is hidden behind the debug util, to enable it, call `wire.app.debug.toggleDetachedCallFeature(true)` and then start/join the call (or just force the component to rerender).

## Screenshots/Screencast (for UI changes)

Desktop wrapper:
Screenshot 2024-04-23 at 14 37 32

Chrome:
Screenshot 2024-04-23 at 14 36 17

Arc (not supporting window popups - it will open in a new tab):
Screenshot 2024-04-23 at 14 34 17

## Checklist

- [x] PR has been self reviewed by the author;
- [x] Hard-to-understand areas of the code have been commented;
- [x] If it is a core feature, unit tests have been added;

### CODE REVIEW

File: src/script/components/DetachedWindow/DetachedWindow.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      `,
+    );
+  }, [height, name, url, width]);
+
+  useEffect(() => {
+    if (!newWindow) {
+      return () => {};
+    }
+
+    setTimeout(() => copyStyles(window.document, newWindow.document), 0);
```
Comment (atomrc) [inline src/script/components/DetachedWindow/DetachedWindow.tsx:65]: isn't there a dom event that we could listen to something like dom ready or something along those lines? 
This `setTimeout` seems a little hacky

File: src/script/components/calling/GroupVideoGridTile.tsx [hunk 2]
```diff
@@ -149,6 +149,7 @@ const GroupVideoGridTile: React.FC<GroupVideoGridTileProps> = ({
           <Video
             autoPlay
             playsInline
+            muted // This is needed to keep playing the video when detached to a new window
```
Comment (atomrc) [inline src/script/components/calling/GroupVideoGridTile.tsx]: could you link some external doc that reference this and the reason why?

Comment (PatrykBuniX) [reply src/script/components/calling/GroupVideoGridTile.tsx]: sure, will do

**Review by atomrc — APPROVED** [review]
A few things to consider, but looking good overall 👌

File: src/script/components/DetachedWindow/DetachedWindow.tsx (hunk 1 above)
Comment (PatrykBuniX) [reply src/script/components/DetachedWindow/DetachedWindow.tsx:65]: I think the new window being opened on a different domain is an issue here, seems like events like this are not fired in this case.

Comment (atomrc) [reply src/script/components/DetachedWindow/DetachedWindow.tsx:65]: Makes sense. Let's add inline doc to this then :)

File: src/script/components/calling/GroupVideoGrid.tsx [hunk 3]
```diff
@@ -164,6 +164,10 @@ const GroupVideoGrid: React.FunctionComponent<GroupVideoGripProps> = ({
             className="group-video__thumbnail-video"
             autoPlay
             playsInline
+            // This is needed to keep playing the video when detached to a new window,
+            // only muted video can be played automatically without user interacting with the window first,
+            // see https://developer.mozilla.org/en-US/docs/Web/Media/Autoplay_guide.
+            muted
```
Comment (atomrc) [inline src/script/components/calling/GroupVideoGrid.tsx]: How about using multi-line comments for this `/* ... */`?

File: src/script/components/calling/GroupVideoGridTile.tsx [hunk 4]
```diff
@@ -149,6 +149,10 @@ const GroupVideoGridTile: React.FC<GroupVideoGridTileProps> = ({
           <Video
             autoPlay
             playsInline
+            // This is needed to keep playing the video when detached to a new window,
+            // only muted video can be played automatically without user interacting with the window first,
+            // see https://developer.mozilla.org/en-US/docs/Web/Media/Autoplay_guide.
```
Comment (atomrc) [inline src/script/components/calling/GroupVideoGridTile.tsx]: How about using multi-line comments for this /* ... */?

File: src/script/components/calling/GroupVideoGrid.tsx (hunk 3 above)
Comment (PatrykBuniX) [reply src/script/components/calling/GroupVideoGrid.tsx]: Sure 👌

**Review by atomrc — APPROVED** [review]
LGTM 😎

**Pull request merged**